add taskrun controller implementation

aaron-prindle opened the pull request:
this pr implements a simple taskrun controller that creates a build and updates the taskrun status to reflect the build status.  currently there are no tests around the controllers and this piece is not fully tested.

tejal29 commented on pkg/apis/pipeline/v1alpha1/taskrun_types.go:
```diff
@@ -88,17 +90,30 @@ type StepRun struct {
+	// AbortedState means prow killed the job early (new commit pushed, perhaps).
+	AbortedState = "aborted"
+	// ErrorState means the job could not schedule (bad config, perhaps).
+	ErrorState = "error"
```
+1 agree with @bobcatfish. Lets add them if we need those and end up using these.

pivotal-nader-ziada replied:
https://docs.google.com/document/d/16j8C91jML4fQRQPhnHihNJUJDcbvW0RM1YAX2REHgyY/edit#heading=h.n8a530nnrb

aaron-prindle replied:
@pivotal-nader-ziada thanks for the`reconcileBuild` links!

I think our plan for Task Status is to adopt the same status conditions as Build and sync the Tasks status to be the same as the Build's status

tejal29 commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -39,6 +44,24 @@ import (
+	groupVersionKind = schema.GroupVersionKind{
+		Group:   v1alpha1.SchemeGroupVersion.Group,
+		Version: v1alpha1.SchemeGroupVersion.Version,
+		Kind:    "TaskRun",
```
add constant.

aaron-prindle replied:
changed groupVersionKind to be a const

tejal29 commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+		case or == nil, or.APIVersion != groupVersionKind.GroupVersion().String(), or.Kind != groupVersionKind.Kind:
+			return nil // Not controlled by this
+		}
+		fmt.Printf("Delete builds/%s", key)
```
use logger here and elsewhere.

tejal29 commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+		}
+		switch or := metav1.GetControllerOf(b); {
+		case or == nil, or.APIVersion != groupVersionKind.GroupVersion().String(), or.Kind != groupVersionKind.Kind:
+			return nil // Not controlled by this
```
`or` i guess is ownerRef

aaron-prindle replied:
or is ownerref which is what is being checked. i've changed the name to be `ownerref`

tejal29 commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	}
+
+	// Ensure taskrun status is correct
+	haveState := tr.Status.State
```
the `ntr` is not used anywhere?
Can we just move lines 219:232 to a function, and use that in L237 in reconcile ?

aaron-prindle replied:
this is an error on my part, ntr should be used in updateStatus.  I will need to move some things around for this

aaron-prindle replied:
I will refactor lines 219:232 to a function

bobcatfish commented:
(plz add link to #59 in commit message)

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	b, err := c.getBuild(namespace, name)
+	switch {
+	case errors.IsNotFound(err):
+		// Do not have a build
```
yes, should return an error

aaron-prindle replied:
if the error is that a build is not found, that means that we need to create a build (haveBuild = false which is it's default value) which is a passthrough here.  In the case that it is any other error (the next switch case) we return an error.  The passthrough is required to catch any other kind of error in the next statement

aaron-prindle replied:
I'll collapse this into one statement

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+
 	// Reconcile this copy of the task run and then write back any status
 	// updates regardless of whether the reconciliation errored out.
 	err = c.reconcile(ctx, tr)
```
I was expecting the actual logic to check/create the build to be in this private func, see example here: https://github.com/knative/serving/blob/76500a6917ce4aec522651677dd9bac677c6034f/pkg/reconciler/v1alpha1/revision/revision.go#L314:22
otherwise consider if we need this func?

bobcatfish replied:
the way that @pivotal-nader-ziada 's example code explicitly expresses phases, then delegates work to them seems pretty slick!

aaron-prindle replied:
yes, I realize now that I put this in the wrong place (Reconcile vs reconcile).  I will move everything to the correct `reconcile` function, thanks for catching this!

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -187,3 +283,70 @@ func (c *Reconciler) updateStatus(taskrun *v1alpha1.TaskRun) (*v1alpha1.TaskRun,
+			},
+		},
+		Spec: *t.Spec.BuildSpec,
+	}, nil
```
you have to check if the BuildSpec has a template and apply that template.
Applying the template includes:
 * checking type of template
 * getting appropriate template object using lister
 * apply the template (use ApplyTemplate() from Build?)

aaron-prindle replied:
I have currently omitted this change as I believe the Build controller will do this for us:
https://github.com/knative/build/blob/e69c21ed74a05d83d68fc15bacbc458807be8288/pkg/controller/build/controller.go#L338

I could definitely be misunderstanding something though, let me know if you still believe we need to call ApplyTemplate()

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
 	// Don't modify the informer's copy.
 	tr := original.DeepCopy()
 
+	// build creation and status update adapted from prow: https://github.com/kubernetes/test-infra/blob/8faa040989cf8552683dcdd4400505ca62a3c408/prow/cmd/build/controller.go#L207
```
There is a tracker defined in the reconciler (meant to track build status) that I don't see used

aaron-prindle replied:
I think I understand now, a tracker is used to monitor child object (in this case a Build) and when the tracker sees a change on the tracked object, it runs `reconcile`?  I have added a tracker.Track statement to reconcile, thanks for pointing out the tracker!

aaron-prindle replied:
@pivotal-nader-ziada 
could you clarify how you thought the tracker would be used, from the comment:
```
	// TODO use build tracker to fetch build
```
it seems that it might be possible to fetch a build with a tracker.  My understanding is that a tracker is an event watcher that runs `reconcile`.  Can you shed some light here?  Thanks!

pivotal-nader-ziada replied:
You are correct Aaron, the tracker is to watch the build. Sorry for the confusing comment

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	switch {
+	case tr.DeletionTimestamp == nil:
+		wantBuild = true
+	}
```
if the `taskRun` is deleted, what is the case you need to create anything? should we just assume its done?

aaron-prindle commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -136,6 +159,79 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	switch {
+	case !wantBuild:
+		if !haveBuild {
+			fmt.Printf("Deleted %s", key)
```
changed, good catch!

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -184,11 +263,48 @@ func (c *Reconciler) updateStatus(taskrun *v1alpha1.TaskRun) (*v1alpha1.TaskRun,
+	}
+	return &buildv1alpha1.Build{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      t.Name,
```
probably want this to be the name of the taskrun, since one task could have many task runs

aaron-prindle replied:
done

bobcatfish commented on pkg/apis/pipeline/v1alpha1/taskrun_types.go:
```diff
@@ -17,17 +17,15 @@ limitations under the License.
 	TaskRef TaskRef     `json:"taskRef"`
 	Trigger TaskTrigger `json:"trigger"`
-	// +optional
+	// +optionalm
```
typo?

aaron-prindle replied:
done

aaron-prindle commented:
all feedback has been addressed

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -44,6 +49,24 @@ const (
+	descMissingCondition = "missing end condition"
+)
+
+const (
```
This should be `var` instead of `const`. Its making the code not compile and thats why the tests in the PR are failing

aaron-prindle replied:
i've changed this back to var

pivotal-nader-ziada commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -159,23 +182,77 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+			return fmt.Errorf("delete build: %v", err)
+		}
+		return nil
+	case tr.Status.Conditions[0].Status != corev1.ConditionUnknown:
```
This will panic if the array is empty which what happened when I tried to run it

aaron-prindle replied:
fixed now, thanks!

pivotal-nader-ziada commented:
@aaron-prindle can you add at least one test to make sure the code runs with no issues, we can add more tests later

aaron-prindle commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -159,23 +182,77 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	switch {
+	case err != nil && !errors.IsNotFound(err):
+		return fmt.Errorf("retrieving build %s for taskRun %s: %v", tr.Name, tr.Name, err)
+	case b.DeletionTimestamp == nil:
```
```
	b, err := c.getBuild(tr.Namespace, tr.Name)
	switch {
       // if there is an error that is build not found
	case err != nil && !errors.IsNotFound(err):
```
this case is for any error that is not "build not found".  if "build is not found", haveBuild will remain false from initialization which is desired.  let me know if i am still missing something or misunderstanding, thanks!

pivotal-nader-ziada replied:
the `b.DeletionTimestamp` will panic if the build is not found and is nil

aaron-prindle replied:
ah of course,  I will change it back to the original form

aaron-prindle commented:
all changes made, changed taskrun.go controller to no longer support delete trigger (mirroring pipelinerun).  this simplifies the controller logic.  also added taskrun_test.go which mirrors pipelinerun_test.go

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -44,6 +49,24 @@ const (
+	descSucceeded        = "succeeded"
+	descFailed           = "failed"
+	descUnknown          = "unknown status"
+	descMissingCondition = "missing end condition"
```
- [ ] i think these can be deleted, doesn't seem like they are used anymore

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -44,6 +49,24 @@ const (
+)
+
+var (
+	groupVersionKind = schema.GroupVersionKind{
```
- [ ] this is misc cleanup but we should change `schema` to `scheme` maybe? kinda weird that we are using both `schema` and `scheme`

aaron-prindle replied:
I'm not sure what is suggested here, the lib's default name is `schema`.  Would you want `schema` to be aliased to something else?

bobcatfish replied:
nope my bad, i misunderstood and thought _we_ had made a package called `schema`

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -141,6 +145,18 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+		Name:       tr.Name,
+	}
+	if err := c.tracker.Track(buildRef, tr); err != nil {
+		logger.Errorf("build %s for taskrun %s: %v", buildRef, tr.Name, err)
```
- [ ] could the error say something about what it was trying to do? e.g. something like `failed to create tracker for build %s for taskrun %s: %v`

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -159,23 +175,47 @@ func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+		// make build obj from task buildspec
+		logger.Infof("make build: %s", tr.Name)
+		if b, err = c.makeBuild(t, tr); err != nil {
+			return fmt.Errorf("make build: %v", err)
```
-  [ ] `failed to create a build for taskrun %s: %v`

bobcatfish commented on pkg/reconciler/v1alpha1/taskrun/taskrun.go:
```diff
@@ -184,11 +224,45 @@ func (c *Reconciler) updateStatus(taskrun *v1alpha1.TaskRun) (*v1alpha1.TaskRun,
+
+func (c *Reconciler) getBuild(namespace, name string) (*buildv1alpha1.Build, error) {
+	return c.buildLister.Builds(namespace).Get(name)
+}
```
- [ ] i dont think we need these wrapper methods, this might have been based on a previous iteration of the pipelinerun controller (for mocking), check out the latest greatness https://github.com/knative/build-pipeline/blob/master/pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go#L69

bobcatfish commented:
One big find here was that the logs in the pods went away too quickly for us to read them, even when trying to read them IMMEDIATELY after the Build completed, so we need a better solution here (@aaron-prindle you or I should make a task in [the october milestone](https://github.com/knative/build-pipeline/milestone/1) for this)
